File: 1/code/4.cpp
#include <iostream>
#include <cmath>

using namespace std;

double intres(int x, double q)
{
    return 4 * M_PI * (x - q / 2 * log((x + q) / x - q));
}

double sumres(int x, double q)
{
    double res = 0;
    for (int i = -x; i <= x; i++)
    {
        // cout << i << endl;
        for (int j = -x; j <= x; j++)
        {
            for (int k = -x; k <= x; k++)
            {
                int n2 = i * i + j * j + k * k;
                if (n2 < x * x)
                {
                    res += 1 / (n2 - q * q);
                }
            }
        }
    }
    return res;
}

double sumreseq(int x, double q)
{
    double res = 0;
    for (int i = -x; i <= x; i++)
    {
        // cout << i << endl;
        for (int j = -x; j <= x; j++)
        {
            for (int k = -x; k <= x; k++)
            {
                int n2 = i * i + j * j + k * k;
                if (n2 == x * x)
                {
                    res += 1 / (n2 - q * q);
                }
            }
        }
    }
    return res;
}

double deltares(int Rl, int Rh, double q)
{
    double res = 0;
    for (int i = -Rh; i <= Rh; i++)
    {
        // cout << i << endl;
        for (int j = -Rh; j <= Rh; j++)
        {
            for (int k = -Rh; k <= Rh; k++)
            {
                int n2 = i * i + j * j + k * k;
                if (n2 < Rh * Rh && n2 >= Rl * Rl)
                {
                    res += 1 / (n2 - q * q);
                }
            }
        }
    }
    return res;
}

int main()
{
    // 从100到2000，输出表面项的差距
    for (int R = 100; R <= 2000; R += 100)
    {
        double f = sumreseq(R, sqrt(0.5));
        cout << f << endl;
    }
}

File: 1/code/4.py
from numpy import arctanh, e, floor, log, pi
from scipy import integrate


def sumres(q):
    res = 0
    for i in range(-R, R+1):
        for j in range(-R, R+1):
            for k in range(-R, R+1):
                n2 = i*i+j*j+k*k
                if n2 <= R*R:
                    res += e**(-(n2-q*q))/(n2-q*q)
    return res


def g(x, q, n2):
    return e**(x*q*q)*(pi/x)**1.5*e**(-pi**2*n2/x)


def intres(q):
    res = 0
    for i in range(-R, R+1):
        for j in range(-R, R+1):
            for k in range(-R, R+1):
                n2 = i*i+j*j+k*k
                if n2 <= R*R and n2 != 0:
                    res += integrate.quad(g, 0, 1, args=(q, n2))[0]
    return res


R = 10
q = 0.5**0.5
# 常数
c = -5.053409213536382
print(sumres(q)+intres(q)+c)
